core/src/main/java/org/apache/calcite/rel/metadata/RelMdPredicates.java

```diff
@@ -666,6 +666,29 @@ static class JoinConditionBasedPredicateInference {
       equivalence = BitSets.closure(equivalence);
     }
 
+    /**
+     * As RexPermuteInputsShuttle, with one exception. When visiting an inputRef,
+     * it will replace the type of the InputRef with the type found in the input fields,
+     * instead of keeping the original type. This is used within
+     * when generating the Left/RightInferredPredicates, to avoid nullability mismatches
+     * between the types of the join and the types of the inputs.
+     */
+    private class TypeChangingRexPermuteInputsShuttle
+        extends org.apache.calcite.rex.RexPermuteInputsShuttle {
+
+      TypeChangingRexPermuteInputsShuttle(Mappings.TargetMapping mapping, RelNode... inputs) {
+        super(mapping, inputs);
+      }
+
+      @Override public RexNode visitInputRef(RexInputRef local) {
+        final int index = local.getIndex();
+        int target = mapping.getTarget(index);
+        return new RexInputRef(
+            target,
+            fields.get(target).getType());
```

Comment:
It would be nice if we were able to check that the original type and the input type are at least compatible, or only change w.r.t. nullability.
That being said, I don't think theres a lot of type checking that happens at the RexNode level.

Reply:
I think in order to check that the two types are equivalent except for nullability, I would need to have access to a typefactory, so I could do something like `typefatory.createTypeWithNullability(t1, t2.isNullable()).equals(t1)`.

For now, I added a sanity check to confirm that the two types are of the same type family. I'm not certain if this is a good check for compatibility, and am open to changing it.